### PR TITLE
Permissions module

### PR DESCRIPTION
**Pull request by mayureshagashe2105**
## Description of the change
Permission module for Corpora resource.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added detailed comments to my code where applicable.
- [x] I have verified that my change does not break existing code.
- [x] My PR is based on the latest changes of the main branch (if unsure, please run `git pull --rebase upstream main`).
- [x] I am familiar with the [Google Style Guide](https://google.github.io/styleguide/) for the language I have coded in.
- [x] I have read through the [Contributing Guide](https://github.com/google/generative-ai-python/blob/main/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: google/generativeai/types/permission_types.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            "name": self.name,
+            "role": self.role,
+            "grantee_type": self.grantee_type,
+            "email_address": self.email_address,
+        }
+
+    @classmethod
+    def get_permission(
```
Comment (MarkDaoust) [inline google/generativeai/types/permission_types.py:190]: nit: or just `Permission.get()`?

File: google/generativeai/types/retriever_types.py [hunk 2]
```diff
@@ -485,6 +488,155 @@ async def list_documents_async(
...
+            `permission_types.Permission` object with specified parent, role, grantee type, and email address.
+
+        Raises:
+            ValueError: When email_address is specified and grantee_type is set to EVERYONE.
+            ValueError: When email_address is not specified and grantee_type is not set to EVERYONE.
+        """
+        if client is None:
+            client = get_dafault_permission_client()
+
+        if (
```
Comment (MarkDaoust) [inline google/generativeai/types/retriever_types.py:538]: nit: these checks could go in `_make_create_permissions_request`.

**Review by MarkDaoust — COMMENTED** [review]
I'm at 4/6 files reviewed, I just have the tests left to look at. Everything looks good so far.

File: tests/test_permission.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                testcase_name="create_permission_failure_email_not_set_when_grantee_type_is_not_everyone",
+                role=2,
+                grantee_type=1,
+                email_address=None,
+            ),
+        ]
+    )
+    def test_create_permission(self, role, grantee_type, email_address):
+        x = retriever.create_corpus("demo-corpus")
+        if (role, grantee_type, email_address) == (2, 3, None):
```
Comment (MarkDaoust) [inline tests/test_permission.py:141]: Using a switch to split the code paths in a parameterized defeats the purpose of parameterized, and makes things less direct. 

Just define three separate tests.

File: tests/test_permission.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                name="corpora/demo_corpus/permissions/123456789",
+                role=3,
+                grantee_type=3,
+            )
+
+    @parameterized.named_parameters(
+        [
+            dict(
+                testcase_name="create_permission_success",
+                role=2,
```
Comment (MarkDaoust) [inline tests/test_permission.py:121]: For human-readability prefer strings to enum-indices [everywhere].

**Review by MarkDaoust — CHANGES_REQUESTED** [review]
Just a couple of tiny fixes and this is good to go.

File: google/generativeai/types/permission_types.py (hunk 1 above)
Comment (mayureshagashe2105) [reply google/generativeai/types/permission_types.py:190]: hmmm...my initial thought was, it being a classmethod `get_permission` should be emphasized.  As the method is not enforced on self just `get` didn't sound right.  On the other hand, considering it is a `Permission`'s method, `get_permission` seems redundant. 

I'll change it to `Permission.get()`  and  will keep its public interface as it is `genai.permission.get_permission()` .

**Review by MarkDaoust — APPROVED** [review]

**Pull request merged**